dgesswein/mfm
# Support 5.10 kernel / Debian bullseye

Hi David,

The following changes make it possible to run a newer Linux kernel and Debian distribution on the BeagleBone Black, Green, and Green Wireless with the MFM Emulator cape.

Updating the image provides many advantages, including but not limited to: better USB device support (including USB and microSD card hotplug functionality, and newer wireless drivers), better desktop support (USB Ethernet and serial support under [modern macOS, Windows and Linux](https://docs.beagleboard.org/latest/intro/support/getting-started.html#troubleshooting)), faster file system performance (I have measured a 30% speed increase using `fio(1)`, modern versions of user utilities, updated SSL/SSH encryption ciphers/root certificates, etc.

The commits in this PR are structured to make the changes easier to understand, or to allow reversion of any specific change if requested. The overall changes include, roughly in order of commits in this PR:

* The kernel-based cape manager is no longer available in Linux 4.x and forward. Instead, a proper MFM Emulator overlay now replaces it, one that supports runtime configuration using the universal overlay's `pin-config` command. This has the added advantage of allowing reconfiguration of the device as reader or emulator, without requiring a reboot.
* The overlay reserves the expected DMA channel and associated PaRAM blocks to prevent conflicts with Linux's eDMA driver.
* `DCHMAP_#` is no longer set correctly by the system for the expected DMA/PaRAM blocks. PRU initialisation code has been updated to include initialisation. These changes are backwards-compatible, effectively acting as a NOP on startup, and have been verified under the existing (3.8-based) image.
* The `setup_emu` and `setup_mfm_read` scripts have been updated to support both the original 3.8.x release as well as newer kernels. For newer kernels, `.bbio` files are provided that configure the pin matrix appropriately. These scripts also correctly set all pins such that switching modes without a reboot is supported on BBB and BBG.
* A write barrier is added to `pru_write_word` to handle the gcc `-Ofast` option, which in turn enables `-fallow-store-data-races`, thus no longer ensuring the write ordering as expected in `pru_exec_cmd` of `PRU0_CMD_DATA` prior to `PRU0_CMD`. This would occasionally prevent `mfm_read` from correctly advancing from RPM detection to drive seek, as the PRU_CMD would be written before the parameter, forcing the PRU incorrectly to seek hundreds of thousands of cylinders. (This can sometimes be seen as an unusually long delay in buffered seek testing.)
* The cape device tree overlay is enhanced to support to `mmap(2)` the shared DDR memory without requiring root access. This takes advantage of the [Linux Kernel Userspace I/O driver](https://www.kernel.org/doc/html/latest/driver-api/uio-howto.html) and `udev(7)` rules to creates a new device at `/dev/uio/prcm` on boot and set permissions accordingly. Additionally the `cap_sys_nice=eip` capabiliity is set on the `mfm_emu` binary at build time to allow re-nicing threads to the real time priority.
* A bug fix to `powerfail` is made to prevent a boot loop should emulation-on-boot be enabled, and the BB is powered, but the cape is not powered (_e.g._, if the BB powered by mini/micro-USB). 
* A new, optional `--sync` flag is added to `mfm_emu`; when specified, this sets `O_DSYNC` on the emulation file, requiring all write operations to the file to complete a full data write (as if a call to `fdatasync(2)` follows every call to `write(2)` to hardware before returning success. This is intended for future experimentation, though early tests show no appreciable performance penalty under moderate load.
* Finally, the `util/` subdirectory includes the necessary files and support to compile and install the overlay, mandatory `modprobe(8)` options, `udev(7)` settings for the UIO devices, and a Debian-specific script to ensure the overlay is rebuilt and installed after any kernel updates via the `apt(8)` subsystem.

A microSD card image based on the official [BeagleBoard Debian image](https://forum.beagleboard.org/t/debian-11-bullseye-testing-images/30542) with these changes applied is available for testing here:

   https://archive.decromancer.ca/mfm/bb_mfm_bullseye_2023-07-11.dd.xz

To test:

* Burn this image (using dd, Balena Etcher, etc.) to a 4GB or larger microSD card, and insert it into a BBB/BBG/BBGW.
* Hold the User button down while powering up. This requires using a chopstick, pencil, or other non-conductive rod to hold the button down while applying power. The button can be released after all 4 LEDs light, then extinguish.
* Login to the image is `debian` with password `temppwd`. (Root remote login is disabled by default.)
* The software resides in `/opt/mfm` (so, `/opt/mfm/emu/mfm_emu` or `/opt/mfm/mfm/mfm_read`).
* Run the programs as the `debian` user; root access is no longer necessary.

This image has been tested by a few people using real hardware, including a PDP-11/23 (RQDX3), an IRIS 2400, and a TI Explorer (custom MFM-SCSI bridge). So far, there are no issues, and performance is improved in some cases.

Please let me know if you have any questions or would like any changes.
